# Исходные данные
Датасет: https://www.openml.org/d/44

Предметная область: спам в электронной рассылке.
Задача: определить, является ли электронное письмо спамом.
Записей: 4601, атрибутов: 57.

## Атрибуты
1.	Частота использования строки «make» (вещественный тип, [0,100])
2.	Частота использования строки «address» (вещественный тип, [0,100])
3.	Частота использования строки «all» (вещественный тип, [0,100])
4.	Частота использования строки «3d» (вещественный тип, [0,100])
5.	Частота использования строки «our» (вещественный тип, [0,100])
6.	Частота использования строки «over» (вещественный тип, [0,100])
7.	Частота использования строки «remove» (вещественный тип, [0,100])
8.	Частота использования строки «internet» (вещественный тип, [0,100])
9.	Частота использования строки «order» (вещественный тип, [0,100])
10.	Частота использования строки «mail» (вещественный тип, [0,100])
11.	Частота использования строки «receive» (вещественный тип, [0,100])
12.	Частота использования строки «will» (вещественный тип, [0,100])
13.	Частота использования строки «people» (вещественный тип, [0,100])
14.	Частота использования строки «report» (вещественный тип, [0,100])
15.	Частота использования строки «addresses» (вещественный тип, [0,100])
16.	Частота использования строки «free» (вещественный тип, [0,100])
17.	Частота использования строки «business» (вещественный тип, [0,100])
18.	Частота использования строки «email» (вещественный тип, [0,100])
19.	Частота использования строки «you» (вещественный тип, [0,100])
20.	Частота использования строки «credit» (вещественный тип, [0,100])
21.	Частота использования строки «your» (вещественный тип, [0,100])
22.	Частота использования строки «font» (вещественный тип, [0,100])
23.	Частота использования строки «000» (вещественный тип, [0,100])
24.	Частота использования строки «money» (вещественный тип, [0,100])
25.	Частота использования строки «hp» (вещественный тип, [0,100])
26.	Частота использования строки «hpl» (вещественный тип, [0,100])
27.	Частота использования строки «george» (вещественный тип, [0,100])
28.	Частота использования строки «650» (вещественный тип, [0,100])
29.	Частота использования строки «lab» (вещественный тип, [0,100])
30.	Частота использования строки «labs» (вещественный тип, [0,100])
31.	Частота использования строки «telnet» (вещественный тип, [0,100])
32.	Частота использования строки «857» (вещественный тип, [0,100])
33.	Частота использования строки «data» (вещественный тип, [0,100])
34.	Частота использования строки «415» (вещественный тип, [0,100])
35.	Частота использования строки «85» (вещественный тип, [0,100])
36.	Частота использования строки «technology» (вещественный тип, [0,100])
37.	Частота использования строки «1999» (вещественный тип, [0,100])
38.	Частота использования строки «parts» (вещественный тип, [0,100])
39.	Частота использования строки «pm» (вещественный тип, [0,100])
40.	Частота использования строки «direct» (вещественный тип, [0,100])
41.	Частота использования строки «cs» (вещественный тип, [0,100])
42.	Частота использования строки «meeting» (вещественный тип, [0,100])
43.	Частота использования строки «original» (вещественный тип, [0,100])
44.	Частота использования строки «project» (вещественный тип, [0,100])
45.	Частота использования строки «re» (вещественный тип, [0,100])
46.	Частота использования строки «edu» (вещественный тип, [0,100])
47.	Частота использования строки «table» (вещественный тип, [0,100])
48.	Частота использования строки «conference» (вещественный тип, [0,100])
49.	Частота использования символа “<” (вещественный тип, [0,100])
50.	Частота использования символа “(” (вещественный тип, [0,100])
51.	Частота использования символа “[” (вещественный тип, [0,100])
52.	Частота использования символа “!” (вещественный тип, [0,100])
53.	Частота использования символа “$” (вещественный тип, [0,100])
54.	Частота использования символа “#” (вещественный тип, [0,100])
55.	Средняя длина непрерывной последовательности заглавных букв (вещественный тип, [1, …])
56.	Самая длинная непрерывная последовательность заглавных букв (целый тип, [1, …])
57.	Сумма длин всех непрерывных последовательностей заглавных букв (целый тип, [1, …])

## Классы
0 – не спам, 1 – спам.

In [1]:
# Импорт библиотек
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

from sklearn.model_selection import train_test_split
from sklearn import svm
from time import time

In [2]:
# Загрузка данных из файла
data = pd.read_csv('data/spambase.data', header = None).values
# Разделение на классы и атрибуты
attributes = data[:, :-1]
classes = np.ravel(data[:, -1:].astype(np.int64, copy=False))
# Выделение обучающей и тестовой выборки
x_train, x_test, y_train, y_test = train_test_split(
    attributes, classes, test_size=0.3, random_state=42)

# Сравнение по ядру

In [3]:
# Ядро Linear
t0 = time()
linear = svm.SVC(kernel="linear").fit(x_train, y_train)
print("Затраченное время:", time() - t0, "c.")
print("Точность ядра Linear: {:.2%}".format(linear.score(x_test, y_test)))

Затраченное время: 554.1157522201538 c.
Точность ядра Linear: 92.76%


In [4]:
# Ядро RBF
t1 = time()
rbf = svm.SVC(kernel="rbf").fit(x_train, y_train)
print("Затраченное время:", time() - t1, "c.")
print("Точность ядра RBF: {:.2%}".format(rbf.score(x_test, y_test)))

Затраченное время: 0.956209659576416 c.
Точность ядра RBF: 81.25%


In [5]:
# Ядро Sigmoid
t2 = time()
sigmoid = svm.SVC(kernel="sigmoid").fit(x_train, y_train)
print("Затраченное время: ", time() - t2, "c.")
print("Точность ядра Sigmoid: {:.2%}".format(sigmoid.score(x_test, y_test)))

Затраченное время:  0.615405797958374 c.
Точность ядра Sigmoid: 35.63%


|            |Linear|RBF  |Sigmoid|
|:-----------|:----:|:---:|:-----:|
|точность (%)|92,76 |81,25|35,63  |
|время (с)   |554   |0,96 |0,62   |

# Сравнение по параметру C

In [7]:
linearC05 = svm.SVC(kernel="linear", C=0.5).fit(x_train, y_train)
rbfC05 = svm.SVC(kernel="rbf", C=0.5).fit(x_train, y_train)
sigmoidC05 = svm.SVC(kernel="sigmoid", C=0.5).fit(x_train, y_train)
print("C = 0.5")
print("Точность ядра Linear: {:.2%}".format(linearC05.score(x_test, y_test)))
print("Точность ядра RBF: {:.2%}".format(rbfC05.score(x_test, y_test)))
print("Точность ядра Sigmoid: {:.2%}".format(sigmoidC05.score(x_test, y_test)))

C = 0.5
Точность ядра Linear: 92.83%
Точность ядра RBF: 78.78%
Точность ядра Sigmoid: 38.52%


In [9]:
linearC10 = svm.SVC(kernel="linear", C=10).fit(x_train, y_train)
rbfC10 = svm.SVC(kernel="rbf", C=10).fit(x_train, y_train)
sigmoidC10 = svm.SVC(kernel="sigmoid", C=10).fit(x_train, y_train)
print("C = 10")
print("Точность ядра Linear: {:.2%}".format(linearC10.score(x_test, y_test)))
print("Точность ядра RBF: {:.2%}".format(rbfC10.score(x_test, y_test)))
print("Точность ядра Sigmoid: {:.2%}".format(sigmoidC10.score(x_test, y_test)))

C = 10
Точность ядра Linear: 91.89%
Точность ядра RBF: 83.13%
Точность ядра Sigmoid: 33.67%


In [10]:
linearC100 = svm.SVC(kernel="linear", C=100).fit(x_train, y_train)
rbfC100 = svm.SVC(kernel="rbf", C=100).fit(x_train, y_train)
sigmoidC100 = svm.SVC(kernel="sigmoid", C=100).fit(x_train, y_train)
print("C = 100")
print("Точность ядра Linear: {:.2%}".format(linearC100.score(x_test, y_test)))
print("Точность ядра RBF: {:.2%}".format(rbfC100.score(x_test, y_test)))
print("Точность ядра Sigmoid: {:.2%}".format(sigmoidC100.score(x_test, y_test)))

C = 100
Точность ядра Linear: 91.38%
Точность ядра RBF: 82.40%
Точность ядра Sigmoid: 33.38%


| Ядро  | C = 0,5 | C = 1 | C = 10 | C = 100 |
|:----- |:-------:|:-----:|:------:|:-------:|
|Linear |92,83    |92,76  |91,89   |91,38    |
|RBF    |78,78    |81,25  |83,13   |82,40    |
|Sigmoid|38,52    |35,63  |33,67   |33,38    |

# Сравнение по параметру gamma ядра RBF

В предыдущих экспериментах наилучшее соотношение между точностью и временем выполнения показало ядро RBF с аргументом С = 10. Точность для данного набора параметров можно улучшить путём подбора коэффициента gamma.

In [19]:
rbfGamma01 = svm.SVC(kernel="rbf", C=10, gamma=0.1).fit(x_train, y_train)
print("Gamma = 0.1:\t{:.2%}".format(rbfGamma01.score(x_test, y_test)))
rbfGamma05 = svm.SVC(kernel="rbf", C=10, gamma=0.5).fit(x_train, y_train)
print("Gamma = 0.5:\t{:.2%}".format(rbfGamma05.score(x_test, y_test)))
rbfGamma001 = svm.SVC(kernel="rbf", C=10, gamma=0.01).fit(x_train, y_train)
print("Gamma = 0.01:\t{:.2%}".format(rbfGamma001.score(x_test, y_test)))
rbfGamma0001 = svm.SVC(kernel="rbf", C=10, gamma=0.001).fit(x_train, y_train)
print("Gamma = 0.001:\t{:.2%}".format(rbfGamma0001.score(x_test, y_test)))
rbfGamma00001 = svm.SVC(kernel="rbf", C=10, gamma=0.0001).fit(x_train, y_train)
print("Gamma = 0.0001:\t{:.2%}".format(rbfGamma00001.score(x_test, y_test)))
rbfGamma0005 = svm.SVC(kernel="rbf", C=10, gamma=0.005).fit(x_train, y_train)
print("Gamma = 0.005:\t{:.2%}".format(rbfGamma0005.score(x_test, y_test)))

Gamma = 0.1:	74.44%
Gamma = 0.5:	70.82%
Gamma = 0.01:	85.59%
Gamma = 0.001:	87.04%
Gamma = 0.0001:	84.29%
Gamma = 0.005:	86.39%


# Вывод

Использование метода опорных векторов (SVM) может быть оправдано для различных наборов данных в зависимости от подобранных параметров. Для данного конкретного набора данных наивысшую точность показало ядро "linear" (92,83%), однако продолжительность времени вычислений (554 секунды или ~9,2 мин) оставляет желать лучшего. Лучший компромисс между точностью и продолжительностью вычислений показало ядро "rbf" (точность ~80%, время выполнения ~1 сек), точность которого была улучшена с изменением параметров C и gamma (лучшая результирующая точность: 87%). Следовательно, можно сделать вывод о наилучшей применимости к задаче классификации для данного набора данных метода опорных векторов с ядром "rbf", C = 10, gamma = 0.001.